src/reader/read_stream.ts

```diff
@@ -0,0 +1,188 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as gax from 'google-gax';
+import * as protos from '../../protos/protos';
+
+import {ReadClient} from './read_client';
+import {logger} from '../util/logger';
+import {Readable, Transform} from 'stream';
+
+type ReadSession = protos.google.cloud.bigquery.storage.v1.IReadSession;
+type ReadRowsResponse =
+  protos.google.cloud.bigquery.storage.v1.IReadRowsResponse;
+
+export type RemoveListener = {
+  off: () => void;
+};
+
+/**
+ * ReadStream is responsible for reading data from a GRPC read stream
+ * connection against the Storage Read API readRows method.
+ *
+ * @class
+ * @extends EventEmitter
+ * @memberof reader
+ */
+export class ReadStream {
+  private _streamName: string;
+  private _offset: number;
+  private _readClient: ReadClient;
+  private _session: ReadSession;
+  private _readStream?: Readable;
+  private _connection?: gax.CancellableStream | null;
+  private _callOptions?: gax.CallOptions;
+
+  constructor(
+    streamName: string,
+    session: ReadSession,
+    readClient: ReadClient,
+    options?: gax.CallOptions
+  ) {
+    this._streamName = streamName;
+    this._session = session;
+    this._offset = 0;
+    this._readClient = readClient;
+    this._callOptions = options;
+    this.open();
+  }
+
+  open() {
+    if (this.isOpen()) {
+      this.close();
+    }
```

Comment:
could this have any unwanted side effects? Or is it just being extra cautious and ensuring it's closed before we call open again?